pkg/credentialprovider/rancher/rancher_registry_credentials.go

```diff
@@ -0,0 +1,128 @@
+package rancher_credentials
+
+import (
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/rancher/go-rancher/client"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+)
+
+// rancher provider
+type rancherProvider struct {
+	credGetter credentialsGetter
+}
+
+// credentials getter from Rancher private registry
+type rancherCredentialsGetter struct {
+	client *client.RancherClient
+}
+
+type rConfig struct {
+	Global configGlobal
+}
+
+// An interface for testing purposes.
+type credentialsGetter interface {
+	getCredentials() []registryCredential
+}
+
+type configGlobal struct {
+	CattleURL       string `gcfg:"cattle-url"`
+	CattleAccessKey string `gcfg:"cattle-access-key"`
+	CattleSecretKey string `gcfg:"cattle-secret-key"`
+}
+
+type registryCredential struct {
+	credential *client.RegistryCredential
+	serverIP   string
+}
+
+func init() {
+	client, err := getRancherClient()
+	if err != nil {
+		glog.Errorf("Failed to get rancher client: %v", err)
+	}
+
+	rancherGetter := &rancherCredentialsGetter{
+		client: client,
+	}
+
+	credentialprovider.RegisterCredentialProvider("rancher-registry-creds",
+		&credentialprovider.CachingDockerConfigProvider{
+			Provider: &rancherProvider{rancherGetter},
+			Lifetime: 30 * time.Minute,
+		})
+}
+
+// Assuming it's always enabled
+func (p *rancherProvider) Enabled() bool {
```

Comment:
You should not be enabled if you fail to detect that you are running on Rancher (e.g. no CATTLE env vars.